Optimize registration of user installed fonts

pvollan opened the pull request:
#### b788bbfd09071496ddd5bda1675fe1a41ff80291
<pre>
Optimize registration of user installed fonts
<a href="https://bugs.webkit.org/show_bug.cgi?id=294168">https://bugs.webkit.org/show_bug.cgi?id=294168</a>
<a href="https://rdar.apple.com/150061783">rdar://150061783</a>

Reviewed by Sihui Liu and Chris Dumez.

Improve performance by only registering user installed fonts with font manager when they are actually being used on the page.

* Source/WTF/wtf/cf/TypeCastsCF.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::cachedFontPlatformData):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::platformFontLookupWithFamily):
(WebCore::FontCache::createFontPlatformData):
(WebCore::userInstalledFontMap):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::sandboxExtensionsForUserInstalledFonts):
(WebKit::addUserInstalledFontURLs):
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAdditionalFonts):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerFontMap):

Canonical link: <a href="https://commits.webkit.org/296142@main">https://commits.webkit.org/296142@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c14581930717aa403172b71bcd3974570c804cff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27091 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112623 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57945 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35591 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81537 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110338 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22002 "Found 2 new test failures: fast/events/ios/select-all-with-existing-selection.html http/tests/geolocation/geolocation-get-current-position-does-not-leak.https.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96807 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61912 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14938 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57389 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/99998 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91372 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115724 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/105956 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34476 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25412 "Found 3 new test failures: inspector/model/remote-object-get-properties.html inspector/runtime/getDisplayableProperties.html inspector/runtime/getProperties.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90577 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90314 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35227 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13013 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30219 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17385 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34397 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39938 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130272 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34143 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35433 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37498 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35804 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,23 @@ static void autoActivateFont(const String& name, CGFloat size)
 {
+    if (auto& fontMap = userInstalledFontMap()) {
+        auto fontKey = family.string().convertToASCIILowercase();
+        auto fontURL = fontMap->getOptional(fontKey);
```
Can use `if (auto fontURL = fontMap->getOptional(fontKey)) {` on the next line.

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,23 @@ static void autoActivateFont(const String& name, CGFloat size)
+        auto fontKey = family.string().convertToASCIILowercase();
+        auto fontURL = fontMap->getOptional(fontKey);
+        if (fontURL) {
+            RELEASE_LOG(Fonts, "FontCache user installed fonts contains family %s with fontURL %s", family.string().utf8().data(), fontURL->string().utf8().data());
```
Don't we want `{public}` here?

pvollan replied:
I think we can switch to `{private}` here.

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,23 @@ static void autoActivateFont(const String& name, CGFloat size)
+        auto fontURL = fontMap->getOptional(fontKey);
+        if (fontURL) {
+            RELEASE_LOG(Fonts, "FontCache user installed fonts contains family %s with fontURL %s", family.string().utf8().data(), fontURL->string().utf8().data());
+            RetainPtr<CFURLRef> cfURL = fontURL->createCFURL();
```
Do we really need `<CFURLRef>` ?

pvollan replied:
Ah, good point, will fix!

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,23 @@ static void autoActivateFont(const String& name, CGFloat size)
+
+            CFErrorRef error = nullptr;
+            if (!CTFontManagerRegisterFontsForURL(cfURL.get(), kCTFontManagerScopeProcess, &error)) {
+                RetainPtr<CFStringRef> descriptionCF = adoptCF(CFErrorCopyDescription(error));
```
Do we really need `<CFStringRef>`?

pvollan replied:
Good point, that should not be needed, will fix!

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,23 @@ static void autoActivateFont(const String& name, CGFloat size)
+            if (!CTFontManagerRegisterFontsForURL(cfURL.get(), kCTFontManagerScopeProcess, &error)) {
+                RetainPtr<CFStringRef> descriptionCF = adoptCF(CFErrorCopyDescription(error));
+                String error(descriptionCF.get());
+                RELEASE_LOG(Fonts, "Could not register font family %s, error %s", family.string().utf8().data(), error.utf8().data());
```
Do we want `{public}` here?

pvollan replied:
I think we can use `{private}` here.

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -990,4 +1008,10 @@ void FontCache::platformReleaseNoncriticalMemory()
 
+std::optional<HashMap<String, URL>>& userInstalledFontMap()
+{
+    static NeverDestroyed<std::optional<HashMap<String, URL>>> fontMap;
```
`MainThreadNeverDestroyed` ?

cdumez commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -990,4 +1008,10 @@ void FontCache::platformReleaseNoncriticalMemory()
 #endif
 }
 
+std::optional<HashMap<String, URL>>& userInstalledFontMap()
```
Returning a std::optional here doesn't seem terribly useful, A default constructed HashMap is just a null pointer, it's cheap to construct and it seems you can check if the HashMap is empty when needed?

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+static void addUserInstalledFontURLs(NSString *path, HashMap<String, URL>& fontURLs, Vector<URL>& sandboxExtensionURLs)
 {
+    RetainPtr pathURL = [NSURL fileURLWithPath:path];
```
Let's use `adoptNS()` and `initFileURLWithPath` instead. We try to avoid autoreleasing whenever possible.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
             URL fontURL(nsFontURL.get());
-            fontURLs.append(fontURL);
-            RELEASE_LOG(Process, "Registering font url %s", fontURL.string().utf8().data());
+            auto fontDescriptors = adoptCF(CTFontManagerCreateFontDescriptorsFromURL((__bridge CFURLRef)nsFontURL.get()));
```
`RetainPtr`

cdumez replied:
Also, I think you can use `bridge_cast()` instead of `(__bridge CFURLRef)`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                RetainPtr fontDescriptor = static_cast<CTFontDescriptorRef>(CFArrayGetValueAtIndex(fontDescriptors.get(), i));
+                if (!fontDescriptor)
+                    continue;
+                auto fontNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontNameAttribute)));
```
`RetainPtr`

cdumez replied:
Can we use `checked_cf_cast` instead of static _cast?

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                if (!fontDescriptor)
+                    continue;
+                auto fontNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontNameAttribute)));
+                auto fontDisplayNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontDisplayNameAttribute)));
```
`RetainPtr`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                    continue;
+                auto fontNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontNameAttribute)));
+                auto fontDisplayNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontDisplayNameAttribute)));
+                auto fontFamilyNameAttribute = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(fontDescriptor.get(), kCTFontFamilyNameAttribute)));
```
`RetainPtr`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                String fontName(fontNameAttribute.get());
+                String fontDisplayName(fontDisplayNameAttribute.get());
+                String fontFamilyName(fontFamilyNameAttribute.get());
+                fontURLs.set(fontName.convertToASCIILowercase(), fontURL);
```
Do we need to use `set()` here and below or would `add()` suffice?

pvollan replied:
I believe `add` is sufficient.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+                String fontFamilyName(fontFamilyNameAttribute.get());
+                fontURLs.set(fontName.convertToASCIILowercase(), fontURL);
+                fontURLs.set(fontFamilyName.convertToASCIILowercase(), fontURL);
+                RELEASE_LOG(Process, "Registering font name %s, display name %s, family name %s,  with URL %s", fontName.utf8().data(), fontDisplayName.utf8().data(), fontFamilyName.utf8().data(), fontURL.string().utf8().data());
```
Do we need `{public}` here?

pvollan replied:
I think we can use {private} here as well.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1473,27 +1507,29 @@ static void addUserInstalledFontURLs(NSString *path, Vector<URL>& fontURLs)
 
-    Vector<URL> fontURLs;
+    if (!m_userInstalledFontURLs) {
+        m_userInstalledFontURLs = { };
```
This looks like a no-op? if we're here then `m_userInstalledFontURLs` is nullopt and will still be nullopt after this assignment.

pvollan replied:
Ah, good catch, will fix!

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1473,27 +1507,29 @@ static void addUserInstalledFontURLs(NSString *path, Vector<URL>& fontURLs)
-        fontURLs.append(URL(url.get()));
+        URL fontURL(url.get());
+        String fontName(nsFontName);
+        m_userInstalledFontURLs->set(fontName, fontURL);
```
Does this really need to use `set()` instead of `add()`?

pvollan replied:
I believe we can use `add` here.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1640,6 +1640,13 @@ static float currentBacklightLevel()
     CTFontManagerRegisterFontURLs((__bridge CFArrayRef)fontURLs.get(), kCTFontManagerScopeProcess, true, blockPtr.get());
 }
 
+void WebProcess::registerFontMap(HashMap<String, URL>&& fontMap, Vector<WebKit::SandboxExtensionHandle>&& sandboxExtensions)
```
`WebKit::` likely unnecessary

cdumez commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -242,5 +242,6 @@ messages -> WebProcess : AuxiliaryProcess WantsAsyncDispatchMessage {
 
 #if PLATFORM(COCOA)
     RegisterAdditionalFonts(struct WebKit::AdditionalFonts fonts)
+    RegisterFontMap(HashMap<String, URL> fonts, Vector<WebKit::SandboxExtensionHandle> sandboxExtensions)
```
`WebKit::` likely unnecessary

pvollan replied:
It looks like we need this in `messages.in` files.

pvollan replied:
Thanks for reviewing!

szewai commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -501,6 +501,7 @@ class WebProcess final : public AuxiliaryProcess {
 
 #if PLATFORM(COCOA)
     void registerAdditionalFonts(AdditionalFonts&&);
+    void registerFontMap(HashMap<String, URL>&&, Vector<SandboxExtension::Handle>&& sandboxExtensions);
```
Should we remove `registerAdditionalFonts` now that it's replaced by `registerFontMap`?

pvollan replied:
That is a very good point! This endpoint is being used by a notification observer in the process pool to dynamically register additional fonts on demand, but we could look into unifying the two paths. Would you be ok with addressing this in a follow-up patch, in order to reduce the scope of this patch?

szewai commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,22 @@ static void autoActivateFont(const String& name, CGFloat size)
+                String error(descriptionCF.get());
+                RELEASE_LOG(Fonts, "Could not register font family %{private}s, error %s", family.string().utf8().data(), error.utf8().data());
+            }
+            fontMap->remove(fontKey);
```
If `CTFontManagerRegisterFontsForURL` is slow, should we also remove the keys with the same URL from `fontMap`?

pvollan replied:
That is a good point, will fix!

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,75 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+            for (CFIndex i = 0; i < CFArrayGetCount(fontDescriptors.get()); ++i) {
+                RetainPtr fontDescriptor = static_cast<CTFontDescriptorRef>(CFArrayGetValueAtIndex(fontDescriptors.get(), i));
+                if (!fontDescriptor)
+                    continue;
```
Should we update sandboxExtensionURLs only if `fontURL` is added to map `fontURLs`? So if the fontURL does not make into the font map the we don't need to create sandbox extension for it (and in this case `sandboxExtensionURLs` should probably be a map)

pvollan replied:
That is a good point! I addressed this in the latest patch (I did keep the sandbox extensions in a vector, since it is only an extension for the top level directory, not the individual font files).

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,80 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
+
+static void addUserInstalledFontURLs(NSString *path, HashMap<String, URL>& fontURLs, Vector<URL>& sandboxExtensionURLs)
 {
+    bool addedFontToMap = false;
```
Boolean variables need a prefix, per WebKit coding style. Maybe `didAddFontToMap`?

pvollan replied:
Fixed, thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1426,41 +1426,80 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
-            RELEASE_LOG(Process, "Registering font url %s", fontURL.string().utf8().data());
+            RetainPtr fontDescriptors = adoptCF(CTFontManagerCreateFontDescriptorsFromURL(bridge_cast(nsFontURL.get())));
+            for (CFIndex i = 0; i < CFArrayGetCount(fontDescriptors.get()); ++i) {
+                RetainPtr fontDescriptor = static_cast<CTFontDescriptorRef>(CFArrayGetValueAtIndex(fontDescriptors.get(), i));
```
`checked_cf_cast`?

pvollan replied:
Good point! I added this in the latest patch, along with adding `WTF_DECLARE_CF_TYPE_TRAIT(CTFontDescriptor)` to `TypeCastsCF.h`.

szewai commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -668,6 +669,24 @@ static void autoActivateFont(const String& name, CGFloat size)
 
 std::unique_ptr<FontPlatformData> FontCache::createFontPlatformData(const FontDescription& fontDescription, const AtomString& family, const FontCreationContext& fontCreationContext, OptionSet<FontLookupOptions> options)
 {
+    if (!userInstalledFontMap().isEmpty()) {
```
If the map is empty then `getOptional` below should return `std::nullopt`, so it seems this check is unnecessary.

pvollan replied:
That is a good point! Although, for performance I think it can make sense to keep the check, since we would otherwise always have to call `family.string().convertToASCIILowercase()`. In most cases, the map will be empty.

Thanks for reviewing!

pvollan committed:
Optimize registration of user installed fonts
https://bugs.webkit.org/show_bug.cgi?id=294168
rdar://150061783

Reviewed by Sihui Liu and Chris Dumez.

Improve performance by only registering user installed fonts with font manager when they are actually being used on the page.

* Source/WTF/wtf/cf/TypeCastsCF.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::cachedFontPlatformData):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::platformFontLookupWithFamily):
(WebCore::FontCache::createFontPlatformData):
(WebCore::userInstalledFontMap):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::sandboxExtensionsForUserInstalledFonts):
(WebKit::addUserInstalledFontURLs):
(WebKit::WebProcessPool::registerUserInstalledFonts):
(WebKit::WebProcessPool::registerAdditionalFonts):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::registerFontMap):

Canonical link: https://commits.webkit.org/296142@main
